### ISSUE TITLE
xla device precision issue

### ISSUE THREAD

**Issue by mars1248** [pytorch/xla#8811]
## ❓ Questions and Help
why this easy test can't pass?
```
import torch
import torch_xla
import torch_xla.core.xla_model as xm
import numpy as np

def compare_linear_outputs(batch_size=32, in_features=64, out_features=128, seed=42, rtol=1e-5, atol=1e-8):
    """
    Compare the output differences of Linear modules running on XLA and CUDA devices
    
    Args:
        batch_size: Input batch size
        in_features: Input feature dimension
        out_features: Output feature dimension
        seed: Random seed to ensure reproducibility
        rtol: Relative tolerance
        atol: Absolute tolerance
    """
    # Set random seed for reproducibility
    torch.manual_seed(seed)
    np.random.seed(seed)
    
    # Create input data in float32 type
    input_data = torch.randn(batch_size, in_features, dtype=torch.float32)
    
    # Create Linear modules with identical parameters, ensuring float32 type
    linear_params = {
        'weight': torch.randn(out_features, in_features, dtype=torch.float32),
        'bias': torch.randn(out_features, dtype=torch.float32)
    }
    
    # Run on CUDA
    if torch.cuda.is_available():
        print("Running CUDA test...")
        cuda_device = torch.device('cuda')
        
        # Create model on CUDA
        cuda_linear = torch.nn.Linear(in_features, out_features, dtype=torch.float32)
        with torch.no_grad():
            cuda_linear.weight.copy_(linear_params['weight'])
            cuda_linear.bias.copy_(linear_params['bias'])
        cuda_linear = cuda_linear.to(cuda_device)
        
        # Run forward pass on CUDA
        cuda_input = input_data.to(cuda_device)
        cuda_output = cuda_linear(cuda_input)
        
        # Move results back to CPU for comparison
        cuda_output_cpu = cuda_output.cpu()
    else:
        print("CUDA not available, skipping CUDA test")
        cuda_output_cpu = None
    
    # Run on XLA
    try:
        print("Running XLA test...")
        xla_device = xm.xla_device()
        
        # Create model on XLA
        xla_linear = torch.nn.Linear(in_features, out_features, dtype=torch.float32)
        with torch.no_grad():
            xla_linear.weight.copy_(linear_params['weight'])
            xla_linear.bias.copy_(linear_params['bias'])
        xla_linear = xla_linear.to(xla_device)
        
        # Run forward pass on XLA
        xla_input = input_data.to(xla_device)
        xla_output = xla_linear(xla_input)
        
        # Move results back to CPU for comparison
        xla_output_cpu = xla_output.cpu()
        
        # Ensure XLA operations are executed
        xm.mark_step()
    except Exception as e:
        print(f"XLA test failed: {e}")
        xla_output_cpu = None
    
    # Compare results using torch.allclose
    if cuda_output_cpu is not None and xla_output_cpu is not None:
        is_close = torch.allclose(cuda_output_cpu, xla_output_cpu, rtol=rtol, atol=atol)
        
        print("\nComparison results:")
        if is_close:
            print(f"✓ XLA and CUDA outputs match within tolerance (rtol={rtol}, atol={atol})")
        else:
            print(f"✗ XLA and CUDA outputs differ beyond tolerance (rtol={rtol}, atol={atol})")
            
            # Calculate and display differences
            abs_diff = torch.abs(cuda_output_cpu - xla_output_cpu)
            max_abs_diff = torch.max(abs_diff).item()
            mean_abs_diff = torch.mean(abs_diff).item()
            
            print(f"Maximum absolute error: {max_abs_diff:.8e}")
            print(f"Mean absolute error: {mean_abs_diff:.8e}")
            
            # Find the position of maximum error
            max_diff_idx = torch.argmax(abs_diff.view(-1))
            row = max_diff_idx // out_features
            col = max_diff_idx % out_features
            print(f"Maximum error position: batch={row}, feature={col}")
            print(f"CUDA value: {cuda_output_cpu[row, col].item():.8f}")
            print(f"XLA value: {xla_output_cpu[row, col].item():.8f}")
    else:
        print("Cannot compare results because at least one test failed")

if __name__ == "__main__":
    # These parameters can be adjusted to test different model and input sizes
    compare_linear_outputs(
        batch_size=64,
        in_features=128,
        out_features=256,
        seed=42,
        rtol=1e-5,  # Relative tolerance
        atol=1e-8   # Absolute tolerance
    )
```
I use cuda 12.4 and xla master branch. H800 device

**Comment by ysiraichi** [comment]
Thank you for filing this issue. I am not sure why there is such a disparity in the results. For reference, I've also compared CPU with CUDA:

```
Comparison results:
✗ CPU and CUDA outputs differ beyond tolerance (rtol=1e-05, atol=1e-08)
Maximum absolute error: 1.90734863e-05
Mean absolute error: 1.81709527e-06
Maximum error position: batch=18, feature=161
CPU value: 34.04414749
CUDA value: 34.04412842
✗ CPU and XLA outputs differ beyond tolerance (rtol=1e-05, atol=1e-08)
Maximum absolute error: 1.74558163e-02
Mean absolute error: 2.66835885e-03
Maximum error position: batch=40, feature=19
CPU value: 2.90465426
XLA value: 2.88719845
✗ CUDA and XLA outputs differ beyond tolerance (rtol=1e-05, atol=1e-08)
Maximum absolute error: 1.74553394e-02
Mean absolute error: 2.66835955e-03
Maximum error position: batch=40, feature=19
CUDA value: 2.90465379
XLA value: 2.88719845
```

Even though both fail, it's clear that XLA has the higher error. It's is odd that we get these results, even though we are just running a linear layer. Maybe it has something to do with the default precision for matrix-multiplication of both PyTorch and PyTorch/XLA.

@amjames @miladm @tengyifei any thoughts on this?

**Comment by ManfeiBai** [comment]
Hi, @mars1248, for CUDA support in PyTorch/XLA, `torch_xla<=2.5.0` has support `CUDA=12.4` with built docker image `us-central1-docker.pkg.dev/tpu-pytorch-releases/docker/xla:r2.5.0_3.11_cuda_12.4`, would you mind try stable version `torch_xla==2.5.0` again?:
```
pip install torch~=2.5.0 'torch_xla[tpu]~=2.5.0' \
  -f https://storage.googleapis.com/libtpu-releases/index.html \
  -f https://storage.googleapis.com/libtpu-wheels/index.html
```

and @ysiraichi is helping our current master GPU support too

**Comment by tengyifei** [comment]
Try

```
torch_xla._XLAC._xla_set_mat_mul_precision('highest')
```

and see if that improves the precision.

There's probably a similar CUDA/GPU config in upstream PyTorch.

**Comment by ysiraichi** [comment]
That works for me. Although they still fail the `allclose` test, the errors are now in the same order of magnitude:

```
✗ CPU and CUDA outputs differ beyond tolerance (rtol=1e-05, atol=1e-08)
Maximum absolute error: 1.90734863e-05
Mean absolute error: 1.81494158e-06
Maximum error position: batch=18, feature=161
CPU value: 34.04414749
CUDA value: 34.04412842

✗ CPU and XLA outputs differ beyond tolerance (rtol=1e-05, atol=1e-08)
Maximum absolute error: 2.00271606e-05
Mean absolute error: 1.76729407e-06
Maximum error position: batch=37, feature=51
CPU value: 6.17083549
XLA value: 6.17081547

✗ CUDA and XLA outputs differ beyond tolerance (rtol=1e-05, atol=1e-08)
Maximum absolute error: 7.62939453e-06
Mean absolute error: 9.40804057e-07
Maximum error position: batch=12, feature=49
CUDA value: -26.93432617
XLA value: -26.93431854
```

@mars1248 let me know if that works for you.

**Comment by mars1248** [comment]
Thank  for your answer, I have also aligned the accuracy, set the flag is OK

**Comment by ysiraichi** [comment]
It does look like XLA errors are larger.

```
✗ CPU and CUDA outputs differ beyond tolerance (rtol=1e-05, atol=1e-08)
Maximum absolute error: 3.12500000e-02
Mean absolute error: 1.92224979e-06
Maximum error position: batch=61, feature=242
CPU value: 7.15625000
CUDA value: 7.12500000

✗ CPU and XLA outputs differ beyond tolerance (rtol=1e-05, atol=1e-08)
Maximum absolute error: 2.50000000e-01
Mean absolute error: 1.30615234e-02
Maximum error position: batch=5, feature=35
CPU value: 32.75000000
XLA value: 32.50000000

✗ CUDA and XLA outputs differ beyond tolerance (rtol=1e-05, atol=1e-08)
Maximum absolute error: 2.50000000e-01
Mean absolute error: 1.30615234e-02
Maximum error position: batch=5, feature=35
CUDA value: 32.75000000
XLA value: 32.50000000
```

One thing I can think of is that PyTorch might be upcasting the inputs to `float32` (#8545).
@miladm @tengyifei Any thoughts?

**Comment by ManfeiBai** [comment]
cc @lsy323, do you happen to have any thought for this bfloat16 precision issue?

**Comment by mars1248** [comment]
@ysiraichi Is there a solution?

**Comment by ysiraichi** [comment]
Sorry. I haven't had the time to investigate this, yet.
That said, one thing I think we could try doing is to set `bias=None`. If the error persists, the problem is in the matrix multiply. In that case, we need to check the current lowering and check if PyTorch is doing something that we aren't.

I won't to be able to do it within this week, but I think I can investigate it next week.

**Comment by tengyifei** [comment]
FWIW I asked some XLA GPU people and they said: "There will be some difference because xla most likely lower the kernel to triton, I.e. they compare two different backends and the execution sequence on gpu will be completely different. in cuda case they do all the steps sequentially. in xla case some steps will be fused with the dot."

So likely the fusion/math differences coupled with bf16 being inherently inaccurate (relative to fp32) could explain the problem.
